File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
void solve(TreeNode* root,int &prev,bool& b){
    if(!root || !b)return;
    solve(root->left,prev,b);
    if(prev!=123 && prev>=root->val){
        b=0;
        return;}
    else
    prev=root->val;

    solve(root->right,prev,b);
}
    bool isValidBST(TreeNode* root) {

        int prev=123;
        bool b=1;
        solve(root,prev,b);
        return b?true:false;
        
    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    boolean solve(TreeNode root,long mn,long mx){
        if(root==null)return true;

        if(root.val<=mn || root.val>=mx){
            return false;
        }

         
    
        return solve(root.left,mn,root.val)&&solve(root.right,root.val,mx);
    }
    public boolean isValidBST(TreeNode root) {
      
        return solve(root,Long.MIN_VALUE,Long.MAX_VALUE);
        
    }
}